lib/rust/profiler-macros/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+            true => enabled_profiler(&snake_ident, ObjIdent),
+            false => disabled_profiler(&snake_ident, ObjIdent),
+        };
+        out.extend(profiler_def);
+        for (j, ChildIdent) in idents[i..].iter().enumerate() {
+            let parent_impl = match enabled_level >= i + j {
+                true => enabled_impl_parent(ObjIdent, ChildIdent),
+                false => disabled_impl_parent(ObjIdent, ChildIdent),
+            };
+            out.extend(parent_impl);
+        }
+    }
+    out
+}
+
+
+
+// ================
+// === profile! ===
+// ================
+
+/// Instruments a function.
+#[proc_macro_attribute]
+pub fn profile(
+    args: proc_macro::TokenStream,
+    ts: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    assert!(args.is_empty(), "#[profile] does not expect any arguments");
+    let mut func = syn::parse_macro_input!(ts as syn::ItemFn);
+    let impl_name = &format!("__profiler_wrapped__{}", func.sig.ident);
+    let impl_ident = syn::Ident::new(impl_name, proc_macro2::Span::call_site());
+    let wrapper_ident = std::mem::replace(&mut func.sig.ident, impl_ident);
+    let wrapper = make_wrapper(wrapper_ident, &func);
+    func.vis = syn::Visibility::Inherited;
+
+    // Emit wrapped function and wrapper..
+    let mut out = proc_macro2::TokenStream::new();
+    wrapper.to_tokens(&mut out);
+    func.to_tokens(&mut out);
+    out.into()
+}
+
+fn make_wrapper(wrapper_ident: syn::Ident, func: &syn::ItemFn) -> proc_macro2::TokenStream {
+    let profiler_label = make_label(&wrapper_ident);
+
+    // Create the wrapper signature, based on the impl signature.
+    let vis = func.vis.clone();
+    let mut wrapper_sig = func.sig.clone();
+    wrapper_sig.ident = wrapper_ident;
+    let impl_ident = &func.sig.ident;
+
+    // Gather argument idents for call in wrapper
+    let args = forward_args(&wrapper_sig.inputs);
+
+    // Parse profiler argument in signature.
+    let last_arg = match wrapper_sig.inputs.last_mut().unwrap() {
+        syn::FnArg::Typed(pat_type) => pat_type,
+        _ => panic!("Last argument to function annotated with #[profile] must be a profiler type"),
+    };
+    let profiler_ident = last_arg.pat.clone();
+
+    // Replace profiler type `T `in sig with `impl Parent<T>`.
+    let profiler_type_in = last_arg.ty.clone();
+    let profiler_generic_parent = quote::quote! { impl profiler::Parent<#profiler_type_in> };
+    last_arg.ty = Box::new(syn::Type::Verbatim(profiler_generic_parent));
+
+    let call = match func.sig.asyncness {
+        None => quote::quote! { #impl_ident(#args) },
+        Some(_) => quote::quote! { #impl_ident(#args).await },
+    };
+    quote::quote! {
+        #vis #wrapper_sig {
+            let _profiler = #profiler_ident.new_child(#profiler_label);
+            let #profiler_ident = _profiler.profiler;
+            #call
+        }
+    }
+}
+
+/// Create a Measurement label, with its file:line info determined by the proc_macro's call site.
```

Comment:
Please use the correct Rust doc syntax when referring to code structures, like ``` [`Measurement`] ```.